plugins/woocommerce/src/Internal/AddressProvider/AddressProviderController.php

```diff
@@ -117,6 +117,22 @@ public function get_registered_providers(): array {
 			$providers[] = $provider_instance;
 		}
 
+			// Check if there's a preferred provider set in the settings.
+		$preferred_provider_id = get_option( 'woocommerce_address_autocomplete_provider', '' );
```

Comment:
I'd wrap this whole block in a if block `! empty( $providers )`:
- execute only of there are providers set up
- keep preferred_provider_id in the if scope
```
if (! empty( $providers )) {
$preferred_provider_id = get_option( 'woocommerce_address_autocomplete_provider', '' );
if ( ! empty( $preferred_provider_id ) ) { ...}

}
```

Reply:
I see you have a `get_preferred_provider` function. Could you reuse it here and test against the empty string to executed your logic when you have a providers? The function already checks for empty providers etc

Reply:
> I see you have a `get_preferred_provider` function. Could you reuse it here and test against the empty string to executed your logic when you have a providers? The function already checks for empty providers etc

`get_preferred_provider` calls `get_registered_providers` so it would be an infinite loop.

Reply:
My thoughts around the two functions are that
- there is some code duplication: 
     - ` get_option('woocommerce_address_autocomplete_provider', '')` is called in both functions. Could we store this somewhere?
     - we check for empty( $providers ) in both places
- the `get_registered_providers` does a couple of things; the sorting based on the preferred provider logic is one of its responsibilities. How do you feel about extracting that in a separate function?
- could we work around the  infinite loop by passing the providers otherwise? This would enable us to have a dedicated function to get the  preferred provider logic, which we could use for the sorting logic.